### ISSUE TITLE
NTP servers not displayed

### ISSUE THREAD

**Issue by j-ed** [nextcloud/serverinfo#160]
## Expected behavior
The NTP servers should be displayed.

## Current behavior
The NTP servers are not displayed.

![grafik](https://user-images.githubusercontent.com/2572764/67198678-c511d100-f3ff-11e9-8f27-f165b67a1ebc.png)

## Steps to reproduce
1. Open the systems app in the settings dialog.

I tried to find the root cause of the problem and identified an incorrect command line syntax causing the problem in `lib/OperatingSystems/DefaultOs.php`:
```
    public function getTimeServers() {
        $servers = shell_exec('cat /etc/ntp.conf |grep  \'^pool\' | cut -f 2 -d " "');
        $servers .= ' ' . shell_exec('cat /etc/systemd/timesyncd.conf |grep  \'^NTP=\' | cut -f 2 -d " "');
        return $servers;
    }
```
This assumes that the pool command is always being used in each `ntp.conf` file what isn't requured. In my configuration file only one server is given, e.g.
```
server <name-of-my-time-server>
```
BTW, I would recommend to change the command syntax as follows, because it
a) handles "pool" and "server" entries at the same time.
b) handles multiple spaces between the keyword and the value.
c) only requires two instead of three commands.
```
$servers = shell_exec('grep -E \'^pool|^server\' /etc/ntp.conf | awk -F" " \'{print $2}\';');
```
Due to the fact that it might be possible that the web server user is not allowed to read the content of `/etc/ntp.conf`, it would be a good idea to print out an error message, e.g. "/etc/ntp.conf permission denied". On my server I had to add the mentioned user to the ntp group to be allowed to read the parameters.

#### Server Configuration
OS: Linux 3.16.70
Web server: Apache2 2.4.41
Database: MariaDB 10.2.24
PHP version: 7.3.10
Nextcloud version: 17.0.0
**Serverinfo app version: 1.7.0**

#### Client Configuration
Browser: Mozilla Firefox 69.0.3
Operating system: Windows 10

**Comment by tflidd** [comment]
If you propose already a fix, create a pull request. We can then review it and merge it.

**Comment by radoeka** [comment]
The fix is going to work for my system, that is using the newer chrony NTP package.

```
# ls -l  /etc/ntp.conf /etc/systemd/timesyncd.conf /etc/chrony.conf
ls: cannot access '/etc/ntp.conf': No such file or directory
-rw-r----- 1 root chrony 1291 Aug 13 07:08 /etc/chrony.conf
-rw-r--r-- 1 root root    615 Jul 31 12:55 /etc/systemd/timesyncd.conf
```

No result for:
`# grep ^NTP /etc/systemd/timesyncd.conf`

The file with the right information is  /etc/chrony.conf

```
# awk '/^pool/ { print $2 }' /etc/chrony.conf
ntp.some-ntp-server.org
ntp2.some-ntp-server.org
ntp3.some-ntp-server.org
pool.ntp.org
```

Would it be possible to provide a simple hook, that can be used to get this kind of information when no result is obtained from the buildin query?

if [ timeservers == empty ]; then
  exec hook defined by user (e.g. owncloud/admin/hooks/timeserver.hook. thus

  if [ exists owncloud/admin/hooks/timeserver.hook ]; then
    exec  owncloud/admin/hooks/timeserver.hook
  fi
fi

**Comment by HigH-HawK** [comment]
Since I have a similar issue, that no time servers are being displayed, I had a look into the DefaultOs.php file and the `getTimeServers()` function. Currently it's using the `shell_exec` command, which is considered insecure by some, so I thought, I could maybe rewrite the function in pure PHP.

Since there are multiple time server applications out there and many people are using different ones, it is hard to cover them all. Also as noted above, many people add time servers to their config in a different way (i.e. some start the line with `server`, others start their line with `pool`), so this needs to be covered as well.

This is my main frame so far:

```php
public function getTimeServers() {
	$ntpConf = "/etc/ntp.conf";
	$timeConf = "/etc/systemd/timesyncd.conf";
	$chronyConf = "/etc/chrony.conf";

	if (is_readable($ntpConf) && filesize($ntpConf) > 16 && !empty(trim(file_get_contents($ntpConf)))) {
		preg_match_all("/^server(.*(?=iburst))|^pool(.*(?=iburst))/m", file_get_contents($ntpConf), $ntpMatches);

		for ($i = 0; $size = count($ntpMatches); $i < $size; ++$i)  {
			$servers .= $ntpMatches['1'][$i];
		}
	}

	if (is_readable($timeConf) && filesize($timeConf) > 16 && !empty(trim(file_get_contents($timeConf)))) {
		preg_match_all("/^NTP=(.*)/", file_get_contents($timeConf), $timeMatches);

		for ($i = 0; $size = count($timeMatches); $i < $size; ++$i)  {
			$servers .= $timeMatches['1'][$i];
		}
	}

	if (is_readable($chronyConf) && filesize($chronyConf) > 16 && !empty(trim(file_get_contents($chronyConf)))) {
		preg_match_all("/^server(.*(?=iburst))|^pool(.*(?=iburst))/m", file_get_contents($chronyConf), $chronyMatches);

		for ($i = 0; $size = count($chronyMatches); $i < $size; ++$i)  {
			$servers .= $chronyMatches['1'][$i];
		}
	}

	return $servers;
}
```

It just needs some fine tuning and a way to fetch the proper time server application, currently being in use.

**Comment by HigH-HawK** [comment]
@MorrisJobke @kesselb I guess this issue can be closed as it's resolved with #243

**Comment by kesselb** [comment]
Thx
